### PR TITLE
Refactor Modal component

### LINKED ISSUE #750: Refactor Modal component
## Problem

Our current `Modal` component is very simplistic and could use some improvements.

## Proposed improvements

- UX:
    - Add a close button in the top right.
    - Close when clicking on the overlay.
- DX:
    - Refactor component design to provide more abstraction in holding the modal's `isOpen` state, eliminating the need hold that state in local variables within consuming components.
    - Add a `title` prop.
    - Pass a `close` function as a slot prop which allows the modal children to close the modal.

## Proposed component design

- Within `App.svelte`, insert:

    ```ts
    setModalManagerInContext();
    ```

- Within any consuming component:

    ```svelte
    <script lang="ts">
      const modal = getModalManager();
    </script>

    <Button on:click={() => { modal.open('addColumn'); }}>Open Modal</Button>

    <Modal
      id="addColumn"
      title="Add Column"
      let:close
    >
      <Button on:click={close}>Cancel</Button>
    </Modal>
    ```

    The modal trigger and content are coupled only by sharing the same `id` string, which should be unique across all `Modal` instances. This decoupling allows the trigger and content to exist far apart from each other in the component tree if needed.

- Within `ModalManager.ts`:

    ```ts
    /**
     * Each Modal instance registers itself with the ModalManager on mount.
     * The ModalManager is then able to orchestrate the opening/closing of
     * modals globally. We can even add support here for multiple open modals
     * if needed later.
     */
    class ModalManager {
      // TODO
    }
    
    const contextKey = {};
    
    export function setModalManagerInContext(): void {
      const modalManager = new ModalManager();
      setContext(contextKey, modalManager);
    }

    export function getModalManager(): ModalManager {
      return getContext(contextKey);
    }
    ```

@pavish what do you think of this plan?

### PR DESCRIPTION
**Pull request by seancolsen**
Fixes #750

Pavish and I have discussed this PR on a call, so it's best that he review it to pick up where we left off.

## Notes

- See the Storybook story Systems/Modal for a place to test out all the different modal types.

## Existing Mathesar modals

- This PR only does the minimum amount necessary to make the existing modal instances function correctly.
- Because the Modal component doesn't rely on the central multiplexer to function properly, it actually didn't change much outside the component library.
- The addition of a close button actually makes the existing Mathesar modals look a bit worse in some cases because the title isn't getting set in the right place. I want to fix that in a subsequent PR.

## Reflections

After I got the modals working well in Storybook, I set forth trying to clean up the Mathesar modals a bit. To my chagrin however, I realized that some of the code I was hoping to simplify with these modal improvements actually was no better off.

For example in this code

```svelte
{#if isAddModalOpen}
  <AddEditSchema bind:isOpen={isAddModalOpen}
    isEditMode={activeSchema !== null} schema={activeSchema}/>
{/if}
```

the `{#if}` seems unnecessary. But I see now that its purpose is to re-mount the `AddEditSchema` component.

Also, I was hoping to have `Modal` contain `AddEditSchema` instead of the other way around, because that's the pattern I've been used to seeing in the past. I see now that there are some benefits to having `AddEditSchema` contain `Modal`.

There are several other small points like this that would be much faster to describe on a call and not worth typing out in here.

Basically, I feel a bit sheepish for coming full circle with this modal design back to a design that's strikingly similar to what we already have. The `Modal.svelte` file is definitely more robust now, but aside from that, this PR feels a bit underwhelming.

## Going forward

I do still want to do some clean up of our existing modals -- and I think I can even address some of the points in the previous section. But I'd like it to be in a subsequent PR. Review/discussion will be more focused. And (more importantly), I'd like to get the toast system merged, and implement #759. With a toast system and a confirmation system, then we can go clean up quite a bit at the same time. For example, some modals have additional state that we can offload to the toast system. And the majority of existing modals can be replaced by the confirmation system. I think it'll be most efficient to do that clean up all in one pass.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by seancolsen** [comment]
@pavish would you like to take an early look at the direction I'm heading here because I ended up changing course somewhat in the midst of implementing this.

I still have more work to do here within `Modal.svelte` to support the props and style we discussed, so don't look too closely at that file.

But in `ModalMultiplexer.ts` and `ModalVisibilityStore.ts` and `Modal.stories.svelte` you can see that I'm using a bit of a different pattern to orchestrate the opening and closing of modals. In the ticket, I proposed using a string as a way to tie the modal trigger to its content. With that approach, I don't like that it's possible for two separate modals to unintentionally collide by having the same identifying string.

**Comment by pavish** [comment]
@seancolsen and I discussed on our call and decided on the following:

1. Avoid circular references from the `ModalMultiplexer` to the `ModalVisibilityStore`.
   - We can separate out the id generation logic into a util.
   - We will have a higher layer entity that makes use of both the `Multiplexer` and `ModalVisibilityStore`. We will only be using this entity outside the internal modal scope.
2. Make the default modal close condition to be only by `button`, and not by `overlay` or `esc`.

**Comment by seancolsen** [comment]
@pavish this is ready for review now.

**Review by pavish — APPROVED** [review]
@seancolsen This PR looks good to me. As mentioned in description, feel free to take up styling issues in a subsequent PR.

I have a few minor comments here, which I think can also be taken care of in the subsequent PR. I'll leave the call to you.

File: mathesar_ui/src/component-library/modal/ModalVisibilityStore.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  update(updater: Updater<boolean>): void {
+    this.activeModalStore.update((openModalId) => {
+      const isCurrentlyVisible = this.id === openModalId;
+      const shouldBecomeVisible = updater(isCurrentlyVisible);
+      if (shouldBecomeVisible) {
+        return this.id;
+      }
+      return undefined;
```
[the remark this answers is not in the document]
Comment (seancolsen) [reply mathesar_ui/src/component-library/modal/ModalVisibilityStore.ts]: Thanks for catching this! Addressed in 929d3008bfdc22752db6f7f542cef9b42586e8eb

File: mathesar_ui/src/component-library/modal/ActiveModalStore.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import type {
+  Subscriber, Unsubscriber, Updater, Writable,
+} from 'svelte/store';
+import { writable } from 'svelte/store';
+
+/**
+ * Stores the ID of the currently opened modal
+ */
+export default class ActiveModalStore implements Writable<number | undefined> {
+  private openModalId = writable<number | undefined>(undefined);
```
[the remark this answers is not in the document]
Comment (seancolsen) [reply mathesar_ui/src/component-library/modal/ActiveModalStore.ts]: Addressed in 1b56cb613fcb45050ea8941d193fc886014af8b9

File: mathesar_ui/src/component-library/modal/Modal.svelte [hunk 3]
```diff
@@ -1,56 +1,82 @@
...
   import type { Size } from '@mathesar-component-library/types';
-  import { portal } from '@mathesar-component-library';
+  import { Button, Icon, portal } from '@mathesar-component-library';
+  import type { ModalCloseAction } from './modal.d';
 
   const dispatch = createEventDispatcher();
 
-  // Additional classes
+  export let isOpen = false;
+  // eslint-disable-next-line no-undef-init
```
[the remark this answers is not in the document]
Comment (seancolsen) [reply mathesar_ui/src/component-library/modal/Modal.svelte]: Addressed in 42c4ff42d6885613ae4e150a29185abfece4b4b0

File: mathesar_ui/src/component-library/modal/ModalVisibilityStore.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ * This store acts somewhat like a radio button interface to the
+ * ActiveModalStore. You can read from this store to determine whether a
+ * specific modal is open. And when you write to this store, that write will get
+ * passed up to the ActiveModalStore which ultimately sets the value, ensuring
+ * that only one modal is open at a time.
+ */
+export default class ModalVisibilityStore implements Writable<boolean> {
+  id: number;
+
+  activeModalStore: ActiveModalStore;
```
[the remark this answers is not in the document]
Comment (seancolsen) [reply mathesar_ui/src/component-library/modal/ModalVisibilityStore.ts]: Addressed in 19a40bb73897f587c610b0cbfdac867b976935f1

**Comment by seancolsen** [comment]
Thanks for the review @pavish . Even though I added a few more commits after your review, I'm setting this PR to auto-merge because my additional changes are very small and should be in-line with your suggestions.

**Pull request merged**